### PR TITLE
Adding timeScale and extent

### PR DESCRIPTION
**Pull request by tgdwyer**
This is my first non-trivial purescript effort so please review and comment (gently :) )!

There's two things I guess could be improved but I don't know how precisely:

I put the time stuff in a new module `Graphics.D3.Time` defined in `time.purs`.  It could also go in `Scale` which would mean we wouldn't need to redefine `unsafeDomain`, `unsafeRange`, etc.  However, there's loads more time stuff to come so I guess it makes sense to separate it now?
What is the correct way to share these `unsafe` definitions between Time and Scale without exposing them outside the library?

Also, I left the `extent` type parameter completely general so it can work with Number or JSDate.  What is the correct way to specify a "this type or that" constraint?

Here's the example I'm working on which uses these added bindings:

```
module Graphics.D3.Examples.UpdateGraph where

import Data.Either
import Data.Traversable
import Data.Date
import Data.Foreign
import Data.Foreign.EasyFFI

import Graphics.D3.Base
import Graphics.D3.Util
import Graphics.D3.Selection
import Graphics.D3.SVG.Axis
import Graphics.D3.Scale
import Graphics.D3.Time
import Graphics.D3.Request

-- | This is a PureScript adaptation of d3noob's example: "Update graph data automatically"
-- | http://bl.ocks.org/d3noob/6bd13f974d6516f3e491

{-

Original JavaScript code:
=========================

-}

type DateAndClose = { date :: JSDate, close :: Number }

coerceDatum :: forall a. a -> D3Eff DateAndClose
coerceDatum = unsafeForeignFunction ["x", ""] "{ date: d3.time.format('%d-%b-%y').parse(x.date), close: Number(x.close) }"

margin = {top: 30, right: 20, bottom: 30, left: 50}
width = 600 - margin.left - margin.right
height = 270 - margin.top - margin.bottom

-- valueline = ffi ["xdata", "ydata", ""] "d3.svg.line().x(xdata).y(ydata);"
close x = x.close
date x = x.date

main = do

  xScale <- timeScale
    .. range [0, width]

  yScale <- linearScale
    .. range [height, 0]

-- Define the axes
  xAxis <- axis
    .. scale xScale
    .. orient "bottom"
    .. ticks 5

  yAxis <- axis
    .. scale yScale
    .. orient "left"
    .. ticks 5

  svg <- rootSelect "body" .. append "svg"
    .. attr "width" (width + margin.left + margin.right)
    .. attr "height" (height + margin.top + margin.bottom)
    .. append "g"
      .. attr "transform" ("translate(" ++ show margin.left ++ "," ++ show margin.top ++ ")")

  csv "data.csv" \(Right array) -> do
    typedData <- traverse coerceDatum array

    xScale ... domain $ extent $ date <$> typedData

    yScale ... domain [0, max close typedData]

    svg ... append "g"
      .. attr "class" "x axis"
      .. attr "transform" ("translate(0," ++ show height ++ ")")
      .. renderAxis xAxis

    svg ... append "g"
      .. attr "class" "y axis"
      .. renderAxis yAxis

    x <- toFunction xScale
    y <- toFunction yScale

    svg ... selectAll "ellipse"
        .. bind typedData
      .. enter .. append "ellipse"
        .. attr "class" "datapoint"
        .. attr' "cx" (x <<< date)
        .. attr' "cy" (y <<< close)
        .. attr "rx" "5"
        .. attr "ry" "5"

    -- let xdata = x <$> (date <$> typedData)
    -- let ydata = y <$> (close <$> typedData)

    --let dd = valueline xdata ydata
    -- svg ... append "path"
    --   .. attr "class" "line"
    --   .. attr "d" dd
```

### CODE REVIEW

**Comment by pelotom** [comment]
This looks like a good start!

I agree that it makes sense to start a new module for time.

As far as where to put the unsafe classes, I'm not sure even Haskell has a good answer for this... I think the usual thing would be to make a module called `Unsafe` and just throw all your junk in there.

For the type of `extent`, what about making a type class like

``` purescript
class Magnitude n
```

with instances

``` purescript
instance Magnitude Number
instance Magnitude JSDate
```

and then the type would be

``` purescript
extent :: forall d. (Magnitude d) => [d] -> [d]
```

**Comment by tgdwyer** [comment]
Commit added, what do you think?  (and BTW do you get notified that the pull request has changed automatically when I push?)

**Comment by pelotom** [comment]
Thanks, I'll take a look tonight. No, I only get notified for comments, so it's good to let me know when you're ready for me to review it :)

**Comment by pelotom** [comment]
So you've added the API for the version of `extent` which doesn't take an accessor, whereas the `min` and `max` here do take accessors. To be uniform in our convention of how we map to [the full D3 API](https://github.com/mbostock/d3/wiki/Arrays#d3_min), and until we can find a way to overload everything in one function per your suggestion in #11, we should probably rename `min` and `max` here to `min'` and `max'`, and also provide accessor-less `min` and `max` versions, as well as an `extent'` which _does_ take an accessor. Does that sound reasonable?

**Comment by tgdwyer** [comment]
How's that?  I'll also send a pull-request to fix the examples

File: src/Util.purs [hunk 1]
```diff
@@ -1,17 +1,41 @@
...
+
+max :: forall d m. (Magnitude m) => [m] -> m
+max = unsafeForeignFunction ["data"] "d3.max(data)"
+
+-- extent takes a data array and returns [min,max]
+-- not restricted to Number, i.e. also works with time
+extent :: forall m. (Magnitude m) => [m] -> [m]
+extent = unsafeForeignFunction ["data"] "d3.extent(data)"
+
+extent' :: forall d m. (Magnitude m) => (d->m) -> [d] -> [m]
```
Comment (pelotom) [inline src/Util.purs:38]: Looks like you forgot to export `extent'`?

**Comment by tgdwyer** [comment]
added extent' export

**Comment by pelotom** [comment]
Just for future reference, it's best to do your work on a topic branch, e.g. `time-scales`. That way when I merge the PR using `--no-ff`, it shows up in the git log with a name suggestive of what was being developed there. Thanks!

**Pull request merged**